File: csharp/sublist/Sublist.cs
public enum SublistType
{
    Equal,
    Unequal,
    Superlist,
    Sublist
}

public static class Sublist
{
    public static SublistType Classify<T>(List<T> list1, List<T> list2)
        where T : IComparable
    {
        if (IsEqual(list1, list2))
            return SublistType.Equal;

        if (IsSublist(list1, list2))
            return SublistType.Sublist;

        if (IsSublist(list2, list1))
            return SublistType.Superlist;

        return SublistType.Unequal;
    }

    static bool IsEqual<T>(List<T> list1, List<T> list2)
    {
        if (list1.Count != list2.Count)
            return false;

        for (int i = 0; i < list1.Count; i++)
        {
            T item1 = list1[i];
            T item2 = list2[i];
            if (item1 == null && item2 != null)
                return false;
            else if (item2 != null && item2 == null)
                return false;
            else if (item1 != null && item2 != null && !item1.Equals(item2))
                return false;
        }
        return true;
    }

    static bool IsSublist<T>(List<T> list1, List<T> list2)
    {
        // This code assumes IsEqual Will have been called previously.

        // Special case empty list is always a sub list.
        if (list1.Count == 0)
            return true;

        int minIndex = 0;
        int index;


        index = list2.IndexOf(list1[0], minIndex);
        while (index != -1 && minIndex < list2.Count)
        {
            minIndex = index + 1;

            // Can you fit list 1 in at this point?
            if (list2.Count >= index + list1.Count)
            {
                // It can fit, is it equal?
                if (IsEqual(list1, list2.GetRange(index, list1.Count)))
                {
                    return true;
                }
            }
            else
            {
                // Doesn't fit so not a sub list.
                return false;
            }

            // try the next occurrence starting with first item of list 1
            index = list2.IndexOf(list1[0], minIndex);
        }
        return false;
    }
}

File: javascript/sublist/sublist.js
//
// This is only a SKELETON file for the 'Sublist' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class List {
  constructor(data = []) {
    this.data = [ ...data ]
  }

  compare(other) {
    // Same length and empty is equal, Same length same data is equal
    if (this.data.length === other.data.length)
    { 
      if (this.data.length === 0)
        return 'EQUAL'
      if (this.data.filter((value, index) => value !== other.data[index]).length === 0)
        return 'EQUAL'
    }

    if (this.data.length === 0 && other.data.length > 0)
      // Empty is sublist of non-empty
      return 'SUBLIST'
    else if (this.data.length > 0 && other.data.length === 0)
      // Non-empty is a superlist of empty
      return 'SUPERLIST'
    else if (this.data.length === other.data.length)
      // Already said they are not equal so if they are the same
      // length they can't be super/sublist either
      return 'UNEQUAL' 
    else if (isSubList(this, other))
      return 'SUBLIST'
    else if (isSubList(other, this))
      return 'SUPERLIST'
    // Nothing left but unequal
    return 'UNEQUAL'
  }
}

// Check if a is a SubList of b
function isSubList(a, b)
{
  let i = -1
  // Find first match to first item in a
  i = b.data.findIndex((value, index) => value === a.data[0] && index > i)
  while (i > -1)
  {
    if (i + a.data.length > b.data.length)
      return false // not enough space
    // Count up the items that don't match between a and b.slice(i)
    let rejectCount = a.data.filter((value, index) => value !== b.data[i + index]).length
    // if no rejects it is a sublist
    if (rejectCount === 0)
      return true
    // Otherwise find the next index that matches the start of a
    i = b.data.findIndex((value, index) => value === a.data[0] && index > i)
  }
  // no match found not a sublist
  return false
}
